First draft of barcode tracking updates

teravest opened the pull request:
This is the first draft of barcode tracking updates
It can be merged as is, then built onto for other updates
insterplatemap.py takes the AG platemap and outputs plateinsert.sql
updatebarcoderesults.py takes a list of resultpdfs and outputs updateresults.sql
Tracking.sql is all the sql changes I made to the database for this.
ag_search.psp has been updated to display the new information

Database Changes
New table PLATE -  currently stores a unique plate id, and the plate name, and the date sequenced
    possible additions are robot_used, reagent, etc
New table PLATE_BARCODE currently stores plate_id and barcode
    possible additions include well id
Barcode had 4 fields added biomass_remaining, sequencing_status, extracted, obsolete
Ag_kit_barcodes had 3 fields added restuls_ready, withdrawn and refunded

teravest committed:
First draft of barcode tracking updates

This is the first draft of barcode tracking updates
It can be merged as is then built onto for other updates
insterplatemap.py takes the AG platemap and outputs plateinsert.sql
updatebarcoderesults.py takes a list of resultpdfs and outputs updateresults.sql
Tracking.sql is all the sql changes I made to the database for this.
ag_search.psp has been updated to display the new information

Database Changes
New table PLATE -  currently stores a unique plate id, and the plate name, and the date sequenced
	possible additions are robot_used, reagent, etc
New table PLATE_BARCODE currently stores plate_id and barcode
	possible additions include well id
Barcode had 4 fields added biomass_remaining, sequencing_status, extracted, obsolete

Ag_kit_barcodes had 3 fields added restuls_ready, withdrawn and refunded

ElDeveloper commented:
I'll try to review this one once #612 is merged so I can deploy this in the dev system.

ElDeveloper commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+script_info['brief_description'] = """Take a plate map and creates db insert
+statements for plate and plate_barcode table"""
+script_info['script_description'] = """Input american gut plate map, sql output file"""
+script_info['script_usage'] = [("Examples:","create db insert statment","%prog -i ag_plate_map.txt -o ag_plate_inserts.sql")]
```
Can you break this line to be 80 columns long?

ElDeveloper commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+__copyright__ = "Copyright 2011, The QIIME project"
+__credits__ = ["Emily TerAvest"]
+__license__ = "GPL"
+__version__ = "1.7.0-dev"
```
Can you change this to 1.0.0-dev?

ElDeveloper commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+from __future__ import division
+
+__author__ = "Emily TerAvest"
+__copyright__ = "Copyright 2011, The QIIME project"
```
Can you change this to be `QIIME-DB` or `QIIME Web App`.

ElDeveloper commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+    platename=''
+    print len(platemaplines)
+    for line in platemaplines:
+        #print line
```
can you remove this line?

ElDeveloper commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+            platename = ''
+
+    sqloutfile.close()
+
```
Can you remove the next few empty lines?

ElDeveloper commented on user_scripts/updatebarcoderesults.py:
```diff
@@ -0,0 +1,60 @@
+        barcodeline = filenamesfile.readline()
+    sqloutfile.close()
+    filenamesfile.close()
+
```
Can you remove these blank lines?

ElDeveloper commented on www/wwwadmin/ag_search.psp:
```diff
@@ -171,12 +171,14 @@ if 'search_term' in form:
 							if len(results) > 0:
-								for ag_kit_barcode_id, ag_kit_id, barcode, sample_date, sample_time, site_sampled, participant_name, environment_sampled, notes in results:
+								for ag_kit_barcode_id, ag_kit_id, barcode, sample_date, sample_time, site_sampled, participant_name, environment_sampled,\
+								 notes,results_ready, withdrawn, refunded in results:
```
The spacing for this line seems odd, is this indented with tabs?

adamrp replied:
Yep, there are some tabs in this file, please convert to 4 spaces

adamrp replied:
Oh... there are TONS of tabs in this file.  Maybe to stay consistent, you should actually convert your lines with 4 spaces to tabs? ugh

ElDeveloper replied:
:+1: on consistency.

teravest replied:
I have 4 space tabs, I was trying to differentiate between a line continuation and the indent for the loop.  I changed it to an indent but it looks really strange to me, more confusing then it was before...

ElDeveloper commented:
Minor comments @teravest but looks good overall. @adamrp would you mind taking a quick pass? I've deployed this on webdev for now

adamrp commented on database/queries/Tracking.sql:
```diff
@@ -0,0 +1,68 @@
+ 
+  CREATE TABLE "QIIME_METADATA"."PLATE_BARCODE" 
+   (	"PLATE_ID" NUMBER(*,0) NOT NULL ENABLE, 
+	"BARCODE" CHAR(9 BYTE) NOT NULL ENABLE,  
```
Are there other places in the DB where barcode is limited to 9 characters?  I think we should make it at least 10, and it might as well even be bigger than that -- we are not hurting for space.  There _may_ be times when want to be able to add "a" and "b" to barcodes to indicate different tips of the same swab.

teravest replied:
In the Barcode table barcode is limited to 9 characters.... no point in changing it here....

adamrp commented on database/queries/Tracking.sql:
```diff
@@ -0,0 +1,68 @@
+update plate set sequence_date = '12/18/2013' where plate_id in (28,33);
+
+
+update barcode set sequencing_status = 'WAITING' where  barcode in (select pb.barcode from plate_barcode pb inner join plate p on pb.plate_id = p.plate_id where p.sequence_date is null);
```
I know this is an SQL file, and PEP8 doesn't apply, but I still think it would be nice to break this line so that the subquery is on the next line :)

ElDeveloper replied:
:+1:

adamrp commented on database/queries/Tracking.sql:
```diff
@@ -0,0 +1,68 @@
+update barcode set sequencing_status = 'WAITING' where  barcode in (select pb.barcode from plate_barcode pb inner join plate p on pb.plate_id = p.plate_id where p.sequence_date is null);
+
+
+
```
maybe add a comment as to why these are obsolete.  Also, why are some of them commented out?

teravest replied:
These were obsolete because we never shipped them.  there was a box of barcodes everyone was afraid to throw way. I tracked them then threw them away.  The commented out ones where not in the database.

adamrp commented on database/queries/Tracking.sql:
```diff
@@ -0,0 +1,68 @@
+--update barcode set obsolete = 'Y' where barcode = '000014056';
+--update barcode set obsolete = 'Y' where barcode = '000014058';
+
+update AG_KIT_BARCODES set refunded = 'Y' where barcode = '000013641';
```
Were there actually more refunded barcodes than just this one?  Then again, I don't think we're going to actually come back to this file and edit it each time we change barcode status or refunded fields.  Maybe a lot of this stuff doesn't need to go into the repo

adamrp commented:
For Tracking.sql, maybe we want to _only_ put in the DDL for the tables?  The other stuff is one-off queries (some of which are just select statements) that we don't plan on maintaining, so it probably doesn't need to go into the repo

adamrp commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+script_info['script_description'] = """Input american gut plate map, sql output file"""
+script_info['script_usage'] = [("Examples:","create db insert statment","%prog -i ag_plate_map.txt -o ag_plate_inserts.sql")]
+script_info['output_description']= "sql file"
+script_info['required_options'] = [\
```
I know this is how the make_qiime_file.py script generates this file, but we need almost _none_ of the line continuation characters that it outputs.  Can you remove them?  I'll change that script so that it doesn't output these and submit a pull on the QIIME repo so we can avoid this in the future.

adamrp replied:
Made a pull request [here](https://github.com/qiime/qiime/pull/1393)

adamrp commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+
+
+def main():
+    option_parser, opts, args =\
```
No need to break this line; instead just append the next line onto this one and it's under 79 characters

ElDeveloper commented on database/queries/Tracking.sql:
```diff
@@ -0,0 +1,68 @@
+
```
Can we also change the name of this file to be `tracking.sql`? Remember that `git mv` is your friend.

teravest replied:
Um I can't do this... 
19:37:39 queries@barcodetracking$ ls 
Tracking.sql                        get_barcode_emails.sql
ag_environmental_metadata.sql               get_successfully_processed_studies.sql
ag_human_metadata.sql                   non_environmental_studies_50_or_more_samples.sql
ag_human_metadata_v2.sql                remove_sample_by_name.sql
19:37:42 queries@barcodetracking$ git mv Tracking.sql tracking.sql
fatal: destination exists, source=database/queries/Tracking.sql, destination=database/queries/tracking.sql

teravest replied:
Got it.  @mjterave told me OSX does not has case sensitive filenames.

adamrp replied:
Probably you have a setting set somewhere so that your shell is not case sensitive.  try this:

``` bash
git mv Tracking.sql temp.sql
git mv temp.sql tracking.sql
```

ElDeveloper replied:
Really? Did not know that, good to know though! :+1:

On Feb 12, 2014, at 7:40 PM, teravest notifications@github.com wrote:

> In database/queries/Tracking.sql:
> 
> > @@ -0,0 +1,68 @@
> > +
> > Got it. @mjterave told me OSX does not has case sensitive filenames.
> 
> —
> Reply to this email directly or view it on GitHub.

adamrp commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+        line = line.split(',')
+        if 'Plate' in line[0] and line[1] != '1':
+            #new plate we only care about the first line
+            platename=line[0]
```
spaces around equal sign

adamrp commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+                    if barcode != '':
+                        try:
+                            barcode = int(barcode)
+                        except:
```
If the line above fails, it will be a ValueError, so might as well use `except ValueError` here

ElDeveloper replied:
:+1:

On Feb 12, 2014, at 5:22 PM, adamrp notifications@github.com wrote:

> In user_scripts/insertplatemap.py:
> 
> > -    for line in platemaplines:
> > -        #print line
> > -        line = line.split(',')
> > -        if 'Plate' in line[0] and line[1] != '1':
> > -            #new plate we only care about the first line
> > -            platename=line[0]
> > -            sqlstmt = "insert into plate (plate_id, plate) values (seq_plate_id.nextval, '%s');\n" % platename
> > -            sqloutfile.write(sqlstmt)
> > -        elif line[0] != '':
> > -            #this line contains barcodes
> > -            if len(line[0]) == 1:
> > -                for barcode in line[1:]:
> > -                    if barcode != '':
> > -                        try:
> > -                            barcode = int(barcode)
> > -                        except:
> >   If the line above fails, it will be a ValueError, so might as well use except ValueError here
> 
> —
> Reply to this email directly or view it on GitHub.

adamrp commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+                        try:
+                            barcode = int(barcode)
+                        except:
+                            continue
```
Maybe print a message so that it doesn't fail silently, like

``` python
print "Barcode %s cannot be converted to an integer." % barcode
```

teravest replied:
The plate map contains BLANK  I want it to fail silently on those.

adamrp replied:
In that case, explicitly test whether or not `barcode == "BLANK"` and, if so, `continue` silently.  But if it something that is not "BLANK" (i.e., it _should_ be a barcode), we do not want that to fail silently.

teravest replied:
The lines start with a letter for the well id.

On Wed, Feb 12, 2014 at 7:51 PM, adamrp notifications@github.com wrote:

> In user_scripts/insertplatemap.py:
> 
> > -        #print line
> > -        line = line.split(',')
> > -        if 'Plate' in line[0] and line[1] != '1':
> > -            #new plate we only care about the first line
> > -            platename=line[0]
> > -            sqlstmt = "insert into plate (plate_id, plate) values (seq_plate_id.nextval, '%s');\n" % platename
> > -            sqloutfile.write(sqlstmt)
> > -        elif line[0] != '':
> > -            #this line contains barcodes
> > -            if len(line[0]) == 1:
> > -                for barcode in line[1:]:
> > -                    if barcode != '':
> > -                        try:
> > -                            barcode = int(barcode)
> > -                        except:
> > -                            continue
> 
> In that case, explicitly test whether or not barcode == "BLANK" and, if
> so, continue silently. But if it something that is not "BLANK" (i.e., it
> _should_ be a barcode), we do not want that to fail silently.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/617/files#r9696211
> .

teravest replied:
The format of the platemap is really inconsistent.  It was all manually
entered so in some places it is BLANK, Blank, XXXX, *BLANK, xxxx

I am making a verbose option

On Wed, Feb 12, 2014 at 8:07 PM, Emily TerAvest ejteravest@gmail.comwrote:

> The lines start with a letter for the well id.
> 
> On Wed, Feb 12, 2014 at 7:51 PM, adamrp notifications@github.com wrote:
> 
> > In user_scripts/insertplatemap.py:
> > 
> > > -        #print line
> > > -        line = line.split(',')
> > > -        if 'Plate' in line[0] and line[1] != '1':
> > > -            #new plate we only care about the first line
> > > -            platename=line[0]
> > > -            sqlstmt = "insert into plate (plate_id, plate) values (seq_plate_id.nextval, '%s');\n" % platename
> > > -            sqloutfile.write(sqlstmt)
> > > -        elif line[0] != '':
> > > -            #this line contains barcodes
> > > -            if len(line[0]) == 1:
> > > -                for barcode in line[1:]:
> > > -                    if barcode != '':
> > > -                        try:
> > > -                            barcode = int(barcode)
> > > -                        except:
> > > -                            continue
> > 
> > In that case, explicitly test whether or not barcode == "BLANK" and, if
> > so, continue silently. But if it something that is not "BLANK" (i.e., it
> > _should_ be a barcode), we do not want that to fail silently.
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/617/files#r9696211
> > .

adamrp commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+                            barcode = int(barcode)
+                        except:
+                            continue
+                        if barcode > 1000:
```
What's the purpose of this check?

teravest replied:
This is to skip the well ids

adamrp commented on user_scripts/insertplatemap.py:
```diff
@@ -0,0 +1,75 @@
+                            continue
+                        if barcode > 1000:
+                            #create an insert statment
+                            sqlstmt = "insert into plate_barcode (plate_id, barcode) select plate_id, '%09d' from plate where plate = '%s';\n" % (barcode, platename)
```
break this line please

adamrp commented on user_scripts/updatebarcoderesults.py:
```diff
@@ -0,0 +1,60 @@
+american gut results for and updates the database to reflect having results"""
+script_info['script_description'] = """Takes a text file of the results of ls 
+the american gut results directory and outputs an sql file to update the database"""
+script_info['script_usage'] = [("Examples:","create database update statements","%prog -i listoffiles.txt -o updateresults.sql")]
```
Break this line too, please

adamrp commented on user_scripts/updatebarcoderesults.py:
```diff
@@ -0,0 +1,60 @@
+the american gut results directory and outputs an sql file to update the database"""
+script_info['script_usage'] = [("Examples:","create database update statements","%prog -i listoffiles.txt -o updateresults.sql")]
+script_info['output_description']= "an sql file to update results read status of db"
+script_info['required_options'] = [\
```
line continuation

adamrp commented on user_scripts/updatebarcoderesults.py:
```diff
@@ -0,0 +1,60 @@
+ make_option('-i','--input_fp',type="existing_filepath",help='the input filepath a list of all the files in the results directory'),\
+ make_option('-o','--output_fp',type="new_filepath",help='the output sqlfile')
+]
+script_info['optional_options'] = [\
```
ditto

adamrp commented on user_scripts/updatebarcoderesults.py:
```diff
@@ -0,0 +1,60 @@
+script_info['output_description']= "an sql file to update results read status of db"
+script_info['required_options'] = [\
+ # Example required option
+ make_option('-i','--input_fp',type="existing_filepath",help='the input filepath a list of all the files in the results directory'),\
```
80 characters

adamrp commented on user_scripts/updatebarcoderesults.py:
```diff
@@ -0,0 +1,60 @@
+ make_option('-o','--output_fp',type="new_filepath",help='the output sqlfile')
+]
+script_info['optional_options'] = [\
+ # Example optional option
```
You can just remove this line and the next one

adamrp commented on user_scripts/updatebarcoderesults.py:
```diff
@@ -0,0 +1,60 @@
+
+
+def main():
+    option_parser, opts, args =\
```
Join next line with this one

adamrp commented on user_scripts/updatebarcoderesults.py:
```diff
@@ -0,0 +1,60 @@
+    option_parser, opts, args =\
+       parse_command_line_parameters(**script_info)
+    filenamesfilename=opts.input_fp
+    filenamesfile = open(filenamesfilename, 'r')
```
lol @ the names of these variables.  Maybe just call it "input_fp" :)

ElDeveloper replied:
Or in this case although it's not a standard, it's useful to suffix these filenames with `_fd` to indicate that these are file descriptors.

adamrp commented on user_scripts/updatebarcoderesults.py:
```diff
@@ -0,0 +1,60 @@
+    filenamesfile = open(filenamesfilename, 'r')
+    sqlfile=opts.output_fp
+    sqloutfile=open(sqlfile, 'w')
+    #barcodeline = barcodeslist.readline()
```
Remove this commented code please

adamrp commented on user_scripts/updatebarcoderesults.py:
```diff
@@ -0,0 +1,60 @@
+        sqlstmt = "update barcode set sequencing_status = 'SUCCESS' where barcode = '%s';\n" % barcode
+        sqloutfile.write(sqlstmt)
+        barcodeline = filenamesfile.readline()
+    sqloutfile.close()
```
Do you need a "commit" here?  Or maybe it's safer to force people to explicitly, manually write out "commit"

adamrp commented:
Mostly style comments, but the functionality looks good!  Can't wait to get this in place.  Do you think you can fix it up in time for our planned deployment tomorrow?

teravest commented:
I will make these changes tonight. Most of these files were just so you guys could see what I did to the database.  For example tracking.sql, plateinsert.sql and resultsupdate.sql   and then insterplatemap.py and 
updatebarcoderesults.py where to show you how 2 of the sql files were created.  I'm not sure that any of those 5 files should actually be in the repo, because they should never be used again, but I wanted you guys to be able to see all what I did to get the information.  Does that make sense?

adamrp commented:
Cool, thanks @teravest.  I understand how you wanted us to see the files, but let's remove them (or at least pare them down significantly) for the actual pull request.

teravest committed:
address comments

change the two .py files although I think that the .py
files and the .sql files should be removed from the
pull request after you guys ok them.

teravest committed:
fixed error handling in insertplatemap.py

ElDeveloper commented:
@teravest looks good and sounds good! At least if you remove them now, these will live in the tree history just in case we ever needed them (for any obscure reason we can think of). @adamrp would you mind merging this after you :+1: it?

teravest committed:
removing files we dont want in the repo

teravest commented:
I just pushed a new commit so ag_search.psp should be the only thing
merged.

On Wed, Feb 12, 2014 at 9:21 PM, Yoshiki Vázquez Baeza <
notifications@github.com> wrote:

> @teravest https://github.com/teravest looks good and sounds good! At
> least if you remove them now, these will live in the tree history just in
> case we ever needed them (for any obscure reason we can think of). @adamrphttps://github.com/adamrpwould you mind merging this after you [image:
> :+1:] it?
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/617#issuecomment-34947207
> .

teravest committed:
remove references to extracted

adamrp commented on www/wwwadmin/ag_search.psp:
```diff
@@ -171,12 +171,14 @@ if 'search_term' in form:
 							if len(results) > 0:
-								for ag_kit_barcode_id, ag_kit_id, barcode, sample_date, sample_time, site_sampled, participant_name, environment_sampled, notes in results:
+								for ag_kit_barcode_id, ag_kit_id, barcode, sample_date, sample_time, site_sampled, participant_name, environment_sampled,\
+									notes,results_ready, withdrawn, refunded in results:
```
Can you indent this one more time?  And then rewrite it in assembly?

ElDeveloper replied:
:+1:

teravest replied:
NO I can't I'm sorry

On Thu, Feb 13, 2014 at 3:01 PM, Yoshiki Vázquez Baeza <
notifications@github.com> wrote:

> In www/wwwadmin/ag_search.psp:
> 
> > ```
> >                             from    ag_kit_barcodes
> >                             where   ag_kit_id = '{0}'""".format(ag_kit_id)
> >                         results = ag_data_access.dynamicMetadataSelect(sql).fetchall()
> >                         if len(results) > 0:
> > ```
> > -                               for ag_kit_barcode_id, ag_kit_id, barcode, sample_date, sample_time, site_sampled, participant_name, environment_sampled, notes in results:
> > -                               for ag_kit_barcode_id, ag_kit_id, barcode, sample_date, sample_time, site_sampled, participant_name, environment_sampled,\
> > -                                   notes,results_ready, withdrawn, refunded in results:
> 
> [image: :+1:]
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/617/files#r9729636
> .

ElDeveloper replied:
You are walking on thin ice @teravest
